[mono] Componentize hot reload

lambdageek opened the pull request:
Move all hot reload code to the hot_reload component.

There are some functional changes:
- The metadata update data (`delta_image`, `method_table_update` fields) moved form `MonoImage` to `BaselineInfo` and `DeltaInfo` structs in the component.  So the locking is different now.
- The `metadata_update_init` EE callback was unused, so it's been removed.
- The `metadata_update_published` runtime callback is always defined, and its work is guarded by a runtime check of the `has_updates` flag.
- The managed `GetApplyUpdateCapabilities` method now calls an icall to find out from the hot reload component if the runtime was started in a way that supports hot reload and returns capabilities appropriately.

There are also some changes to the console sample:
- use the hot reload BuildTool instead of the hotreload-delta-gen CLI tool to make deltas
- make it run on osx-arm64 by adhoc codesigning the published binary (other console samples likely need this, too)

lambdageek committed:
Move most of the code to hot_reload.c

For the fast path data, we will leave a struct in the runtime with the values
and pass its address to the component to fill in.

lambdageek committed:
Mark functions used by hot reload with MONO_COMPONENT_API

lambdageek committed:
Drop the metadata_update_init EE and runtime callbacks

They're unused now

lambdageek committed:
remove mono_ prefix from some static methods in hot reload

lambdageek committed:
Always define metadata_update_published runtime callback

lambdageek committed:
Move metadata update data out of MonoImage, into hot_reload component

Use separate baseline and delta structs for tracking updates

lambdageek committed:
move slowpath metadata lookup, and cleanup to hot_reload component

Move the delta_image and method_table_update data into BaselineInfo
and DeltaInfo.  Keep a single `MonoImage:has_updates` flag to switch
over to the slow path through the component.

lambdageek committed:
fixup: set has_updates on first update, too

Also setup the BaselineInfo and DeltaInfo earlier before the first
string heap lookup

lambdageek committed:
fixup after rebase

lambdageek committed:
ConsoleDelta sample: build on osx-arm64, use hotreload buildtool

instead of hotreload-delta-gen CLI tool

lambdageek committed:
fix whitespace

lambdageek committed:
bugfix - get the value of MonoImage:has_updates not its address

lambdageek committed:
Dynamically query runtime for GetApplyUpdateCapabilities

if the hot reload component is stubbed out, return empty capabilities.

This is a step toward removing FEATURE_METADATA_UPDATE and
ENABLE_METADATA_UPDATE defines and solely using component
infrastrucutre for controlling behavior

lambdageek committed:
Mark some runtime functions with MONO_COMPONENT_API

lambdageek committed:
fixup icall declaration

lateralusX commented on src/mono/mono/component/hot_reload.h:
```diff
@@ -7,13 +7,28 @@
+	void (*effective_table_slow) (const MonoTableInfo **t, int *idx);
+	int (*relative_delta_index) (MonoImage *image_dmeta, int token);
 	void (*apply_changes) (MonoImage *base_image, gconstpointer dmeta, uint32_t dmeta_len, gconstpointer dil, uint32_t dil_len, MonoError *error);
+        void (*image_close_except_pools_all) (MonoImage *base_image);
```
nit: Align indention

lateralusX commented on src/mono/mono/metadata/metadata-internals.h:
```diff
@@ -797,6 +787,13 @@ mono_install_image_loader (const MonoImageLoader *loader);
+	int has_updates;
+};
+
+
```
nit: extra new line.

lateralusX commented on src/mono/mono/metadata/metadata-update.c:
```diff
@@ -10,1153 +10,120 @@
-	int index = g_list_index (base->delta_image, delta);
-	g_assert (index != -1);
-	return 1 + index;
+	MonoComponentHotReload *comp = mono_component_hot_reload ();
```
Should this be mono_component_hot_reload ()->init() ?

lambdageek replied:
I don't have an init method.  Do you think I should make one?  `mono_component_hot_reload_init` already does some initialization earlier.  I think having two functions with `init` in the name might be more confusing.

lateralusX replied:
I was confused by the diff, it looked like you had a method just calling into mono_component_hot_reload()->xxx, similar pattern as use elsewhere in this file, but this one called set_fastpath_data, but when I looked in the full file, is see that the intent was not just to have something calling into a hot_reload component method, but actually doing some initialization work as part of mono_metadata_update_init (that I didn't see in diff).

lateralusX reviewed (approved):
Looked primarily on the component implementation, not that deeply into the bulk of moved code. Couple of smaller things, otherwise, LGTM.

lambdageek committed:
fix whitespace and indentation

lambdageek committed:
Return capabilities if hot reload component is available, even if env var isn't set.

Match CoreCLR behavior.  Also makes our testsuite actually run the
tests (we check for capabilities before running setting
DOTNET_MODIFIABLE_ASSEMBLIES using the remote executor)
